File: apps/server/src/projections.ts
/**
 * Continuous speculative merge — projections.
 *
 * A projection for (repo, branch_ref, target_ref) holds the always-up-to-date
 * result of running the Tier 1 merge model over those two branches.
 *
 * The projection is computed LAZILY: on first access or when marked stale.
 * A trigger on the `refs` table marks any matching projection as stale
 * whenever either input ref advances (see migration 0009__projections.sql).
 *
 * Whitepaper §6.7, engineering-plan §7.6.
 */
import type postgres from 'postgres';
import {
  decodeTree,
  encodeCommit,
  encodeTree,
  hashObject,
  type HashAlgo,
  type TreeEntry,
} from '@sharp/git-canonical';
import { tier1Merge, type FileSet } from '@sharp/client/src/merge/index';
import { getObject, putObject, type Sql } from './cas';
import { resolveRef } from './refs';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type ProjectionStatus = 'clean' | 'dilemma' | 'stale' | 'error';

export interface ProjectionRow {
  repo_id: string;
  branch_ref: string;
  target_ref: string;
  branch_tip: string | null;
  target_tip: string | null;
  projection_commit: string | null;
  status: ProjectionStatus;
  dilemma: unknown | null;
  computed_at: Date;
}

interface RawRow {
  repo_id: string;
  branch_ref: string;
  target_ref: string;
  branch_tip: Buffer | null;
  target_tip: Buffer | null;
  projection_commit: Buffer | null;
  status: ProjectionStatus;
  dilemma: unknown | null;
  computed_at: Date;
}

function rowToProjection(raw: RawRow): ProjectionRow {
  return {
    repo_id: raw.repo_id,
    branch_ref: raw.branch_ref,
    target_ref: raw.target_ref,
    branch_tip: raw.branch_tip ? raw.branch_tip.toString('hex') : null,
    target_tip: raw.target_tip ? raw.target_tip.toString('hex') : null,
    projection_commit: raw.projection_commit ? raw.projection_commit.toString('hex') : null,
    status: raw.status,
    dilemma: raw.dilemma,
    computed_at: raw.computed_at,
  };
}

// ---------------------------------------------------------------------------
// CRUD helpers
// ---------------------------------------------------------------------------

export async function registerProjection(
  sql: Sql,
  repo_id: string,
  branch_ref: string,
  target_ref: string,
): Promise<ProjectionRow> {
  const rows = await sql<RawRow[]>`
    insert into projections (repo_id, branch_ref, target_ref, status)
    values (${repo_id}::uuid, ${branch_ref}, ${target_ref}, 'stale')
    on conflict (repo_id, branch_ref, target_ref) do update
      set status = 'stale', computed_at = now()
    returning
      repo_id::text as repo_id, branch_ref, target_ref,
      branch_tip, target_tip, projection_commit,
      status, dilemma, computed_at
  `;
  return rowToProjection(rows[0]!);
}

export async function listProjections(
  sql: Sql,
  repo_id: string,
  status?: ProjectionStatus,
): Promise<ProjectionRow[]> {
  const rows = status
    ? await sql<RawRow[]>`
        select
          repo_id::text as repo_id, branch_ref, target_ref,
          branch_tip, target_tip, projection_commit,
          status, dilemma, computed_at
        from projections
        where repo_id = ${repo_id}::uuid and status = ${status}
        order by branch_ref, target_ref
      `
    : await sql<RawRow[]>`
        select
          repo_id::text as repo_id, branch_ref, target_ref,
          branch_tip, target_tip, projection_commit,
          status, dilemma, computed_at
        from projections
        where repo_id = ${repo_id}::uuid
        order by branch_ref, target_ref
      `;
  return rows.map(rowToProjection);
}

export async function deleteProjection(
  sql: Sql,
  repo_id: string,
  branch_ref: string,
  target_ref: string,
): Promise<void> {
  await sql`
    delete from projections
    where repo_id = ${repo_id}::uuid
      and branch_ref = ${branch_ref}
      and target_ref = ${target_ref}
  `;
}

// ---------------------------------------------------------------------------
// Lazy-compute entry point
// ---------------------------------------------------------------------------

/**
 * Return the projection row, recomputing if stale or missing.
 * Returns undefined if the projection does not exist yet.
 */
export async function getProjection(
  sql: postgres.Sql,
  repo_id: string,
  branch_ref: string,
  target_ref: string,
): Promise<ProjectionRow | undefined> {
  const rows = await sql<RawRow[]>`
    select
      repo_id::text as repo_id, branch_ref, target_ref,
      branch_tip, target_tip, projection_commit,
      status, dilemma, computed_at
    from projections
    where repo_id = ${repo_id}::uuid
      and branch_ref = ${branch_ref}
      and target_ref = ${target_ref}
    limit 1
  `;
  if (rows.length === 0) return undefined;
  const row = rows[0]!;
  if (row.status === 'stale') {
    return recomputeProjection(sql, repo_id, branch_ref, target_ref);
  }
  return rowToProjection(row);
}

// ---------------------------------------------------------------------------
// Recompute
// ---------------------------------------------------------------------------

/**
 * Materialize a commit's tree recursively into a Map<path, content>.
 * Blob contents are decoded as UTF-8; binary files get an empty string
 * (not meaningful for merge but avoids crashing on non-text paths).
 */
async function materializeTree(
  sql: Sql,
  repo_id: string,
  treeId: Uint8Array,
  prefix: string,
  algo: HashAlgo,
): Promise<Map<string, string>> {
  const out = new Map<string, string>();
  const rec = await getObject(sql, repo_id, treeId);
  if (!rec || rec.kind !== 'tree') return out;
  for (const entry of decodeTree(rec.data, algo)) {
    const fullPath = prefix === '' ? entry.name : `${prefix}/${entry.name}`;
    if (entry.mode === '40000') {
      const sub = await materializeTree(sql, repo_id, entry.id, fullPath, algo);
      for (const [p, c] of sub) out.set(p, c);
    } else {
      const blobRec = await getObject(sql, repo_id, entry.id);
      if (blobRec) {
        try {
          out.set(fullPath, Buffer.from(blobRec.data).toString('utf8'));
        } catch {
          out.set(fullPath, '');
        }
      }
    }
  }
  return out;
}

/**
 * Write a Map<string,string> back into the CAS as a tree object.
 * Flat tree for v1 — nested directories are not needed here because
 * materializeTree already flattens paths (e.g. "src/foo.ts").
 * Files are stored as blobs; the tree entries reference them.
 */
async function writeTreeToCas(
  sql: Sql,
  repo_id: string,
  files: Map<string, string>,
  algo: HashAlgo,
): Promise<Uint8Array> {
  const entries: TreeEntry[] = [];
  // Sort names so the tree hash is deterministic.
  const sortedPaths = [...files.keys()].sort();
  for (const path of sortedPaths) {
    const content = files.get(path)!;
    const payload = new Uint8Array(Buffer.from(content, 'utf8'));
    const blobId = await putObject(sql, { repo: repo_id, algo, kind: 'blob', payload });
    // Flat tree: only the basename.
    // For nested paths we need to build sub-trees. We flatten via a
    // recursive helper below instead.
    entries.push({ mode: '100644', name: path, id: blobId });
  }
  const treeBytes = encodeTree(entries, algo);
  const treeId = await putObject(sql, { repo: repo_id, algo, kind: 'tree', payload: treeBytes });
  return treeId;
}

/**
 * Get the commit's tree id by decoding the commit object.
 */
async function getCommitTree(
  sql: Sql,
  repo_id: string,
  commitId: Uint8Array,
): Promise<Uint8Array | undefined> {
  const rec = await getObject(sql, repo_id, commitId);
  if (!rec || rec.kind !== 'commit') return undefined;
  const text = Buffer.from(rec.data).toString('utf8');
  const m = /^tree ([0-9a-f]+)$/m.exec(text);
  return m ? new Uint8Array(Buffer.from(m[1]!, 'hex')) : undefined;
}

const SHARP_COMMITTER = {
  nameAndEmail: 'Sharp <sharp-projection@sharp.dev>',
  timestamp: 0,
  timezone: '+0000',
};

/**
 * Run the merge engine on current tips and update the projections row.
 * Always returns the updated row.
 */
export async function recomputeProjection(
  sql: postgres.Sql,
  repo_id: string,
  branch_ref: string,
  target_ref: string,
): Promise<ProjectionRow> {
  // Resolve current tips.
  const branchTip = await resolveRef(sql, repo_id, branch_ref);
  const targetTip = await resolveRef(sql, repo_id, target_ref);

  if (!branchTip || !targetTip) {
    const rows = await sql<RawRow[]>`
      update projections
      set status = 'error',
          computed_at = now()
      where repo_id = ${repo_id}::uuid
        and branch_ref = ${branch_ref}
        and target_ref = ${target_ref}
      returning
        repo_id::text as repo_id, branch_ref, target_ref,
        branch_tip, target_tip, projection_commit,
        status, dilemma, computed_at
    `;
    return rowToProjection(rows[0]!);
  }

  // Materialize trees. We use target_ref as the "base" (common ancestor
  // approximation for v1; a real LCA walk is a Phase 19 follow-up) and
  // branch_ref as A, target_ref as B.
  const algo: HashAlgo = 'sha1';

  const branchTreeId = await getCommitTree(sql, repo_id, branchTip);
  const targetTreeId = await getCommitTree(sql, repo_id, targetTip);

  if (!branchTreeId || !targetTreeId) {
    const rows = await sql<RawRow[]>`
      update projections
      set status = 'error',
          computed_at = now()
      where repo_id = ${repo_id}::uuid
        and branch_ref = ${branch_ref}
        and target_ref = ${target_ref}
      returning
        repo_id::text as repo_id, branch_ref, target_ref,
        branch_tip, target_tip, projection_commit,
        status, dilemma, computed_at
    `;
    return rowToProjection(rows[0]!);
  }

  const branchFiles = await materializeTree(sql, repo_id, branchTreeId, '', algo);
  const targetFiles = await materializeTree(sql, repo_id, targetTreeId, '', algo);

  // For v1, use target as the base (so we see what branch adds on top of target).
  const base: FileSet = { files: targetFiles };
  const a: FileSet = { files: branchFiles };
  const b: FileSet = { files: targetFiles };

  const mergeResult = await tier1Merge(base, a, b);

  if (mergeResult.outcome === 'clean_ok') {
    // Write the merged tree to CAS and create a projection commit.
    const mergedFiles = mergeResult.files!;

    // timestamp based on current time for the projection commit
    const now = Math.floor(Date.now() / 1000);
    const committer = { ...SHARP_COMMITTER, timestamp: now };

    const mergedTreeId = await writeTreeToCas(sql, repo_id, mergedFiles, algo);

    const commitBytes = encodeCommit(
      {
        tree: mergedTreeId,
        parents: [branchTip, targetTip],
        author: committer,
        committer,
        message: `sharp: projection of ${branch_ref} onto ${target_ref}\n`,
      },
      algo,
    );
    const commitId = hashObject('commit', commitBytes, algo);
    await putObject(sql, { repo: repo_id, algo, kind: 'commit', payload: commitBytes });

    const rows = await sql<RawRow[]>`
      update projections
      set status = 'clean',
          branch_tip = ${Buffer.from(branchTip)},
          target_tip = ${Buffer.from(targetTip)},
          projection_commit = ${Buffer.from(commitId)},
          dilemma = null,
          computed_at = now()
      where repo_id = ${repo_id}::uuid
        and branch_ref = ${branch_ref}
        and target_ref = ${target_ref}
      returning
        repo_id::text as repo_id, branch_ref, target_ref,
        branch_tip, target_tip, projection_commit,
        status, dilemma, computed_at
    `;
    return rowToProjection(rows[0]!);
  }

  if (mergeResult.outcome === 'dilemma') {
    const rows = await sql<RawRow[]>`
      update projections
      set status = 'dilemma',
          branch_tip = ${Buffer.from(branchTip)},
          target_tip = ${Buffer.from(targetTip)},
          projection_commit = null,
          dilemma = ${JSON.stringify(mergeResult.dilemma)}::jsonb,
          computed_at = now()
      where repo_id = ${repo_id}::uuid
        and branch_ref = ${branch_ref}
        and target_ref = ${target_ref}
      returning
        repo_id::text as repo_id, branch_ref, target_ref,
        branch_tip, target_tip, projection_commit,
        status, dilemma, computed_at
    `;
    return rowToProjection(rows[0]!);
  }

  // 'unhandled' — store as error with reason in dilemma field.
  const rows = await sql<RawRow[]>`
    update projections
    set status = 'error',
        branch_tip = ${Buffer.from(branchTip)},
        target_tip = ${Buffer.from(targetTip)},
        projection_commit = null,
        dilemma = ${JSON.stringify({ reason: mergeResult.reason ?? 'unhandled' })}::jsonb,
        computed_at = now()
    where repo_id = ${repo_id}::uuid
      and branch_ref = ${branch_ref}
      and target_ref = ${target_ref}
    returning
      repo_id::text as repo_id, branch_ref, target_ref,
      branch_tip, target_tip, projection_commit,
      status, dilemma, computed_at
  `;
  return rowToProjection(rows[0]!);
}

File: crates/sharp/src/projections.rs
//! Continuous speculative merge — projections.
//!
//! Ports `sharp/apps/server/src/projections.ts`.
//!
//! A projection for `(repo_id, branch_ref, target_ref)` holds the
//! always-up-to-date result of running the Tier-1 (text) merge over those two
//! branches. It is computed *lazily*: on first access or when marked stale. A
//! trigger on `sharp.refs` (see migration `0007_sharp_projections.sql`) marks
//! any matching projection stale whenever either input ref advances.
//!
//! # Port notes
//!
//! - **Commit model:** the TS prototype writes git-canonical two-parent commits
//!   plus standalone tree objects. The Rust crate uses the existing single-parent
//!   JSON commit model ([`crate::commit`]) backed by `sharp.commit_paths` (no
//!   separate tree objects). [`materialize_tree`] therefore reconstructs the
//!   path→content map from `commit_paths` rather than walking tree objects, and
//!   projection commits are stored as JSON commit objects with the branch tip as
//!   their single parent.
//! - **Hashes** are hex `String`s throughout (per `docs/rust-reorg-decisions.md`
//!   §2), not `bytea`.
//! - **Merge:** the per-path combine uses the text Tier-1 merge
//!   ([`crate::semantic_merge::three_way_merge`]); `target_ref` is the base
//!   (common-ancestor approximation for v1, a real LCA walk is a follow-up).
//!
//! See `docs/architecture.md §Sharp — Tier-1 Rust Semantic Merge § Components (crates/sharp)` and whitepaper §6.7.

use std::collections::BTreeMap;

use serde_json::Value as Json;
use sqlx::{PgPool, Row};
use uuid::Uuid;

use crate::error::SharpError;
use crate::object::{self, ObjectType};
use crate::refs;
use crate::semantic_merge::three_way_merge;

// ───────────────────────────────────────────────────────────────────────────────
// Types
// ───────────────────────────────────────────────────────────────────────────────

/// Lifecycle state of a projection row.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProjectionStatus {
    /// Inputs merged cleanly; `projection_commit` holds the merged commit.
    Clean,
    /// The merge surfaced a dilemma; `dilemma` holds the payload.
    Dilemma,
    /// The projection is out of date and must be recomputed on next access.
    Stale,
    /// Recompute failed (e.g. a ref did not resolve); `dilemma` holds the reason.
    Error,
}

impl ProjectionStatus {
    /// The database string for this status.
    pub fn as_str(self) -> &'static str {
        match self {
            ProjectionStatus::Clean => "clean",
            ProjectionStatus::Dilemma => "dilemma",
            ProjectionStatus::Stale => "stale",
            ProjectionStatus::Error => "error",
        }
    }

    /// Parse a status from its database string.
    fn parse_str(s: &str) -> Result<Self, SharpError> {
        match s {
            "clean" => Ok(ProjectionStatus::Clean),
            "dilemma" => Ok(ProjectionStatus::Dilemma),
            "stale" => Ok(ProjectionStatus::Stale),
            "error" => Ok(ProjectionStatus::Error),
            other => Err(SharpError::GitInterop(format!(
                "projection row has unknown status: {other}"
            ))),
        }
    }
}

/// A projection row, mirroring `sharp.projections`.
#[derive(Debug, Clone)]
pub struct ProjectionRow {
    pub repo_id: Uuid,
    pub branch_ref: String,
    pub target_ref: String,
    /// Hex SHA of the branch tip at compute time, if computed.
    pub branch_tip: Option<String>,
    /// Hex SHA of the target tip at compute time, if computed.
    pub target_tip: Option<String>,
    /// Hex SHA of the merged projection commit, set when `status == Clean`.
    pub projection_commit: Option<String>,
    pub status: ProjectionStatus,
    /// Dilemma / error payload, set when `status` is `Dilemma` or `Error`.
    pub dilemma: Option<Json>,
    pub computed_at: chrono::DateTime<chrono::Utc>,
}

/// The full `RETURNING` / `SELECT` column list shared by every query below.
const PROJECTION_COLS: &str = "repo_id, branch_ref, target_ref, branch_tip, target_tip, \
     projection_commit, status, dilemma, computed_at";

/// Map a [`sqlx`] row to a [`ProjectionRow`].
fn row_to_projection(r: &sqlx::postgres::PgRow) -> Result<ProjectionRow, SharpError> {
    let status: String = r.try_get("status")?;
    Ok(ProjectionRow {
        repo_id: r.try_get("repo_id")?,
        branch_ref: r.try_get("branch_ref")?,
        target_ref: r.try_get("target_ref")?,
        branch_tip: r.try_get("branch_tip")?,
        target_tip: r.try_get("target_tip")?,
        projection_commit: r.try_get("projection_commit")?,
        status: ProjectionStatus::parse_str(&status)?,
        dilemma: r.try_get("dilemma")?,
        computed_at: r.try_get("computed_at")?,
    })
}

/// Identity used as author/committer for projection commits.
const SHARP_PROJECTION_AUTHOR: &str = "Sharp <sharp-projection@sharp.dev>";

// ───────────────────────────────────────────────────────────────────────────────
// CRUD helpers
// ───────────────────────────────────────────────────────────────────────────────

/// Register a projection for `(repo_id, branch_ref, target_ref)`, or reset an
/// existing one to `stale`. Returns the (stale) row; call [`recompute_projection`]
/// or [`get_projection`] to materialize it.
///
/// # Errors
///
/// Returns [`SharpError::Db`] on a database error.
pub async fn register_projection(
    pool: &PgPool,
    repo_id: Uuid,
    branch_ref: &str,
    target_ref: &str,
) -> Result<ProjectionRow, SharpError> {
    let sql = format!(
        r#"
        INSERT INTO sharp.projections (repo_id, branch_ref, target_ref, status)
        VALUES ($1, $2, $3, 'stale')
        ON CONFLICT (repo_id, branch_ref, target_ref) DO UPDATE
            SET status = 'stale', computed_at = now()
        RETURNING {PROJECTION_COLS}
        "#
    );
    let row = sqlx::query(&sql)
        .bind(repo_id)
        .bind(branch_ref)
        .bind(target_ref)
        .fetch_one(pool)
        .await?;
    row_to_projection(&row)
}

/// List projections in a repo, optionally filtered by `status`, ordered by
/// `(branch_ref, target_ref)`.
///
/// # Errors
///
/// Returns [`SharpError::Db`] on a database error.
pub async fn list_projections(
    pool: &PgPool,
    repo_id: Uuid,
    status: Option<ProjectionStatus>,
) -> Result<Vec<ProjectionRow>, SharpError> {
    let rows = match status {
        Some(s) => {
            let sql = format!(
                r#"
                SELECT {PROJECTION_COLS}
                FROM   sharp.projections
                WHERE  repo_id = $1 AND status = $2
                ORDER  BY branch_ref, target_ref
                "#
            );
            sqlx::query(&sql)
                .bind(repo_id)
                .bind(s.as_str())
                .fetch_all(pool)
                .await?
        }
        None => {
            let sql = format!(
                r#"
                SELECT {PROJECTION_COLS}
                FROM   sharp.projections
                WHERE  repo_id = $1
                ORDER  BY branch_ref, target_ref
                "#
            );
            sqlx::query(&sql).bind(repo_id).fetch_all(pool).await?
        }
    };
    rows.iter().map(row_to_projection).collect()
}

/// Delete a projection. A no-op if it does not exist.
///
/// # Errors
///
/// Returns [`SharpError::Db`] on a database error.
pub async fn delete_projection(
    pool: &PgPool,
    repo_id: Uuid,
    branch_ref: &str,
    target_ref: &str,
) -> Result<(), SharpError> {
    sqlx::query(
        r#"
        DELETE FROM sharp.projections
        WHERE repo_id = $1 AND branch_ref = $2 AND target_ref = $3
        "#,
    )
    .bind(repo_id)
    .bind(branch_ref)
    .bind(target_ref)
    .execute(pool)
    .await?;
    Ok(())
}

/// Return the projection row, recomputing it if stale. Returns `None` if no
/// projection has been registered for `(repo_id, branch_ref, target_ref)`.
///
/// # Errors
///
/// Returns [`SharpError::Db`] on a database error.
pub async fn get_projection(
    pool: &PgPool,
    repo_id: Uuid,
    branch_ref: &str,
    target_ref: &str,
) -> Result<Option<ProjectionRow>, SharpError> {
    let sql = format!(
        r#"
        SELECT {PROJECTION_COLS}
        FROM   sharp.projections
        WHERE  repo_id = $1 AND branch_ref = $2 AND target_ref = $3
        LIMIT  1
        "#
    );
    let row = sqlx::query(&sql)
        .bind(repo_id)
        .bind(branch_ref)
        .bind(target_ref)
        .fetch_optional(pool)
        .await?;

    let Some(row) = row else { return Ok(None) };
    let projection = row_to_projection(&row)?;
    if projection.status == ProjectionStatus::Stale {
        return recompute_projection(pool, repo_id, branch_ref, target_ref)
            .await
            .map(Some);
    }
    Ok(Some(projection))
}

// ───────────────────────────────────────────────────────────────────────────────
// Materialization
// ───────────────────────────────────────────────────────────────────────────────

/// Materialize a commit's file tree into a `path → content` map.
///
/// Reconstructs the tree from `sharp.commit_paths` (the single-parent JSON
/// commit model; no separate tree objects). Each non-deletion row's blob is
/// loaded and decoded as UTF-8; deletion rows (`blob_sha IS NULL`) are skipped.
/// Binary blobs that are not valid UTF-8 are decoded lossily, matching the TS
/// prototype's "non-text paths do not crash the merge" behavior.
///
/// # Errors
///
/// Returns [`SharpError::Db`] on a database error or [`SharpError::ObjectNotFound`]
/// if a referenced blob is missing.
pub async fn materialize_tree(
    pool: &PgPool,
    repo_id: Uuid,
    commit_sha: &str,
) -> Result<BTreeMap<String, String>, SharpError> {
    let rows = sqlx::query(
        r#"
        SELECT path, blob_sha
        FROM   sharp.commit_paths
        WHERE  repo_id = $1 AND commit_sha = $2
        ORDER  BY path
        "#,
    )
    .bind(repo_id)
    .bind(commit_sha)
    .fetch_all(pool)
    .await?;

    let mut out: BTreeMap<String, String> = BTreeMap::new();
    for r in &rows {
        let path: String = r.try_get("path")?;
        let blob_sha: Option<String> = r.try_get("blob_sha")?;
        let Some(blob_sha) = blob_sha else { continue };
        let data = object::load(pool, &blob_sha).await?;
        out.insert(path, String::from_utf8_lossy(&data).into_owned());
    }
    Ok(out)
}

// ───────────────────────────────────────────────────────────────────────────────
// Recompute
// ───────────────────────────────────────────────────────────────────────────────

/// Combine `branch` over `target` per path using the text Tier-1 merge.
///
/// `target` is treated as the merge base (v1 LCA approximation). For each path:
/// the base is its `target` content (or empty if absent), `ours` is the `branch`
/// content (or empty), `theirs` is the `target` content (or empty). The result
/// is the union of all paths. A path present in `branch` but absent in `target`
/// is taken verbatim from `branch`; a `target`-only path is taken verbatim.
fn combine_text(
    branch: &BTreeMap<String, String>,
    target: &BTreeMap<String, String>,
) -> BTreeMap<String, String> {
    let mut merged: BTreeMap<String, String> = BTreeMap::new();
    for path in branch.keys().chain(target.keys()) {
        if merged.contains_key(path) {
            continue;
        }
        match (branch.get(path), target.get(path)) {
            // Present on both sides → 3-way merge with target as the base.
            (Some(ours), Some(theirs)) => {
                merged.insert(path.clone(), three_way_merge(theirs, ours, theirs));
            }
            // Branch-only or target-only → take the present side verbatim. A
            // path missing on the other side is an addition, not a deletion;
            // running the line merge would read the absent side as an empty
            // file and drop the content.
            (Some(only), None) | (None, Some(only)) => {
                merged.insert(path.clone(), only.clone());
            }
            (None, None) => unreachable!("path came from one of the two maps"),
        }
    }
    merged
}

/// Whether any merged file still carries conflict markers, indicating the text
/// merge could not resolve cleanly (the projection's "dilemma" condition).
fn has_conflict(files: &BTreeMap<String, String>) -> bool {
    files
        .values()
        .any(|c| c.contains("<<<<<<< ours") || c.contains(">>>>>>> theirs"))
}

/// Write a `path → content` map as a single-parent JSON projection commit.
///
/// Stores each file's content as a blob, records the blobs as `commit_paths`,
/// and writes a JSON commit object (the same shape [`crate::commit::commit`]
/// produces) with `parent` set to `branch_tip`. Unlike [`crate::commit::commit`],
/// this does **not** move any branch ref — a projection is not a branch and must
/// not retrigger the stale trigger. Returns the projection commit SHA.
async fn write_projection_commit(
    pool: &PgPool,
    repo_id: Uuid,
    branch_tip: &str,
    branch_ref: &str,
    target_ref: &str,
    files: &BTreeMap<String, String>,
) -> Result<String, SharpError> {
    // Store blobs and collect (path, blob_sha) pairs.
    let mut paths: Vec<(String, String)> = Vec::with_capacity(files.len());
    for (path, content) in files {
        let blob_sha = object::store(pool, repo_id, ObjectType::Blob, content.as_bytes()).await?;
        paths.push((path.clone(), blob_sha));
    }

    let message = format!("sharp: projection of {branch_ref} onto {target_ref}\n");
    let payload = serde_json::json!({
        "tree":    Json::Null,
        "parent":  branch_tip,
        "message": message,
        "author":  SHARP_PROJECTION_AUTHOR,
    });
    let data =
        serde_json::to_vec(&payload).expect("serde_json::to_vec never fails for a simple Value");
    let commit_sha = object::store(pool, repo_id, ObjectType::Commit, &data).await?;

    sqlx::query(
        r#"
        INSERT INTO sharp.commit_metadata
            (commit_sha, repo_id, parent_sha, message, author, authored_at)
        VALUES ($1, $2, $3, $4, $5, now())
        ON CONFLICT (commit_sha) DO NOTHING
        "#,
    )
    .bind(&commit_sha)
    .bind(repo_id)
    .bind(branch_tip)
    .bind(&message)
    .bind(SHARP_PROJECTION_AUTHOR)
    .execute(pool)
    .await?;

    for (path, blob_sha) in &paths {
        sqlx::query(
            r#"
            INSERT INTO sharp.commit_paths (commit_sha, repo_id, path, blob_sha)
            VALUES ($1, $2, $3, $4)
            "#,
        )
        .bind(&commit_sha)
        .bind(repo_id)
        .bind(path)
        .bind(blob_sha)
        .execute(pool)
        .await?;
    }

    Ok(commit_sha)
}

/// Update a projection row to `error` with no resolved tips (a ref did not
/// resolve). Returns the updated row.
async fn store_error(
    pool: &PgPool,
    repo_id: Uuid,
    branch_ref: &str,
    target_ref: &str,
    reason: &str,
) -> Result<ProjectionRow, SharpError> {
    let sql = format!(
        r#"
        UPDATE sharp.projections
           SET status = 'error',
               dilemma = $4,
               computed_at = now()
         WHERE repo_id = $1 AND branch_ref = $2 AND target_ref = $3
        RETURNING {PROJECTION_COLS}
        "#
    );
    let row = sqlx::query(&sql)
        .bind(repo_id)
        .bind(branch_ref)
        .bind(target_ref)
        .bind(serde_json::json!({ "reason": reason }))
        .fetch_one(pool)
        .await?;
    row_to_projection(&row)
}

/// Recompute the projection on the current ref tips and persist the result.
///
/// Resolves `branch_ref` and `target_ref` to their current tips, materializes
/// each commit's file tree, combines `branch` over `target` with the text Tier-1
/// merge, and updates the row:
///
/// - both refs resolve and the merge is conflict-free → `clean`, with a fresh
///   projection commit;
/// - the merge leaves conflict markers → `dilemma`, no commit;
/// - either ref fails to resolve → `error`.
///
/// Always returns the updated row.
///
/// # Errors
///
/// Returns [`SharpError::Db`] on a database error or [`SharpError::ObjectNotFound`]
/// if a referenced blob is missing.
pub async fn recompute_projection(
    pool: &PgPool,
    repo_id: Uuid,
    branch_ref: &str,
    target_ref: &str,
) -> Result<ProjectionRow, SharpError> {
    let branch_tip = refs::resolve_ref(pool, repo_id, branch_ref).await?;
    let target_tip = refs::resolve_ref(pool, repo_id, target_ref).await?;

    let (Some(branch_tip), Some(target_tip)) = (branch_tip, target_tip) else {
        return store_error(
            pool,
            repo_id,
            branch_ref,
            target_ref,
            "branch_ref or target_ref did not resolve",
        )
        .await;
    };

    let branch_files = materialize_tree(pool, repo_id, &branch_tip).await?;
    let target_files = materialize_tree(pool, repo_id, &target_tip).await?;

    let merged = combine_text(&branch_files, &target_files);

    if has_conflict(&merged) {
        let involved: Vec<&String> = merged
            .iter()
            .filter(|(_, c)| c.contains("<<<<<<< ours") || c.contains(">>>>>>> theirs"))
            .map(|(p, _)| p)
            .collect();
        let dilemma = serde_json::json!({
            "sharp_dilemma": 1,
            "kind": "dilemma",
            "reason": "text merge produced conflicts",
            "involved_paths": involved,
        });
        let sql = format!(
            r#"
            UPDATE sharp.projections
               SET status = 'dilemma',
                   branch_tip = $4,
                   target_tip = $5,
                   projection_commit = NULL,
                   dilemma = $6,
                   computed_at = now()
             WHERE repo_id = $1 AND branch_ref = $2 AND target_ref = $3
            RETURNING {PROJECTION_COLS}
            "#
        );
        let row = sqlx::query(&sql)
            .bind(repo_id)
            .bind(branch_ref)
            .bind(target_ref)
            .bind(&branch_tip)
            .bind(&target_tip)
            .bind(dilemma)
            .fetch_one(pool)
            .await?;
        return row_to_projection(&row);
    }

    let commit_sha =
        write_projection_commit(pool, repo_id, &branch_tip, branch_ref, target_ref, &merged)
            .await?;

    let sql = format!(
        r#"
        UPDATE sharp.projections
           SET status = 'clean',
               branch_tip = $4,
               target_tip = $5,
               projection_commit = $6,
               dilemma = NULL,
               computed_at = now()
         WHERE repo_id = $1 AND branch_ref = $2 AND target_ref = $3
        RETURNING {PROJECTION_COLS}
        "#
    );
    let row = sqlx::query(&sql)
        .bind(repo_id)
        .bind(branch_ref)
        .bind(target_ref)
        .bind(&branch_tip)
        .bind(&target_tip)
        .bind(&commit_sha)
        .fetch_one(pool)
        .await?;
    row_to_projection(&row)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn status_round_trips() {
        for s in [
            ProjectionStatus::Clean,
            ProjectionStatus::Dilemma,
            ProjectionStatus::Stale,
            ProjectionStatus::Error,
        ] {
            assert_eq!(ProjectionStatus::parse_str(s.as_str()).unwrap(), s);
        }
    }

    #[test]
    fn status_unknown_errors() {
        assert!(ProjectionStatus::parse_str("bogus").is_err());
    }

    #[test]
    fn combine_takes_branch_only_path_verbatim() {
        let mut branch = BTreeMap::new();
        branch.insert("a.rs".to_string(), "fn a() {}\n".to_string());
        let target = BTreeMap::new();
        let merged = combine_text(&branch, &target);
        assert_eq!(merged.get("a.rs").map(String::as_str), Some("fn a() {}\n"));
    }

    #[test]
    fn combine_takes_target_only_path_verbatim() {
        let branch = BTreeMap::new();
        let mut target = BTreeMap::new();
        target.insert("b.rs".to_string(), "fn b() {}\n".to_string());
        let merged = combine_text(&branch, &target);
        assert_eq!(merged.get("b.rs").map(String::as_str), Some("fn b() {}\n"));
    }

    #[test]
    fn combine_clean_when_only_branch_changed() {
        // base == theirs == target, ours == branch differs → take branch.
        let mut branch = BTreeMap::new();
        branch.insert("f.rs".to_string(), "v2\n".to_string());
        let mut target = BTreeMap::new();
        target.insert("f.rs".to_string(), "v1\n".to_string());
        let merged = combine_text(&branch, &target);
        // base = target ("v1"), ours = branch ("v2"), theirs = target ("v1")
        // → base == theirs → take ours.
        assert_eq!(merged.get("f.rs").map(String::as_str), Some("v2\n"));
        assert!(!has_conflict(&merged));
    }

    #[test]
    fn has_conflict_detects_markers() {
        let mut files = BTreeMap::new();
        files.insert(
            "c.rs".to_string(),
            "<<<<<<< ours\na\n=======\nb\n>>>>>>> theirs\n".to_string(),
        );
        assert!(has_conflict(&files));
    }

    #[test]
    fn has_conflict_false_for_clean() {
        let mut files = BTreeMap::new();
        files.insert("c.rs".to_string(), "clean\n".to_string());
        assert!(!has_conflict(&files));
    }
}
